File: python/mundane/inspect.py
"""Read-only inspection helpers (don't take the write lock)."""

from __future__ import annotations

import base64
import json
import sqlite3
from typing import Any, Dict, List

from ._schema import SCHEMA_VERSION


def _open_ro(path: str) -> sqlite3.Connection:
    return sqlite3.connect(f"file:{path}?mode=ro", uri=True)


def _check_schema(conn: sqlite3.Connection) -> None:
    row = conn.execute(
        "SELECT value FROM mundane_meta WHERE key = 'schema_version'"
    ).fetchone()
    if not row or row[0] != SCHEMA_VERSION:
        raise RuntimeError(
            f"unsupported schema version: {row[0] if row else None!r}"
        )


def status(path: str) -> Dict[str, Any]:
    """Return a summary dict for the task at `path`."""
    conn = _open_ro(path)
    try:
        _check_schema(conn)
        meta = dict(conn.execute("SELECT key, value FROM mundane_meta").fetchall())
        counts = conn.execute(
            "SELECT status, COUNT(*) FROM mundane_steps GROUP BY status"
        ).fetchall()
        by_status = {s: c for s, c in counts}
        total = sum(by_status.values())
        return {
            "path": path,
            "task_id": meta.get("task_id"),
            "created_at": meta.get("created_at"),
            "total_steps": total,
            "done": by_status.get("done", 0),
            "pending": by_status.get("pending", 0),
            "failed": by_status.get("failed", 0),
        }
    finally:
        conn.close()


def steps(path: str) -> List[Dict[str, Any]]:
    """Return a list of step rows ordered by id."""
    conn = _open_ro(path)
    try:
        _check_schema(conn)
        rows = conn.execute(
            "SELECT id, name, kind, encoding, status, started_at, finished_at, error "
            "FROM mundane_steps ORDER BY id"
        ).fetchall()
        return [
            {
                "id": r[0], "name": r[1], "kind": r[2], "encoding": r[3],
                "status": r[4], "started_at": r[5], "finished_at": r[6],
                "error": r[7],
            }
            for r in rows
        ]
    finally:
        conn.close()


def get_result(path: str, name: str) -> Any:
    """Return the decoded cached result for the step named `name`."""
    conn = _open_ro(path)
    try:
        _check_schema(conn)
        row = conn.execute(
            "SELECT encoding, result, status FROM mundane_steps WHERE name = ?",
            (name,),
        ).fetchone()
        if row is None:
            raise KeyError(f"no step named {name!r}")
        enc, result, st = row
        if st != "done":
            raise RuntimeError(f"step {name!r} is {st}, not done")
        if result is None:
            return None
        # SQLite may give us bytes or str depending on storage; normalise.
        text = result.decode("utf-8") if isinstance(result, bytes) else result
        if enc == "json":
            return json.loads(text)
        if enc == "text":
            return text
        if enc == "b64":
            return base64.b64decode(text)
        if enc == "epoch":
            return int(text)
        raise RuntimeError(f"unknown encoding: {enc!r}")
    finally:
        conn.close()

File: go/mundane/inspect.go
package mundane

import (
	"database/sql"
	"fmt"
)

type Status struct {
	Path      string
	TaskID    string
	CreatedAt string
	Total     int
	Done      int
	Pending   int
	Failed    int
}

type InspectStepRow struct {
	ID         int64
	Name       string
	Kind       string
	Encoding   string
	Status     string
	StartedAt  string
	FinishedAt string
	Error      string
}

func openRO(path string) (*sql.DB, error) {
	db, err := sql.Open("sqlite", "file:"+path+"?mode=ro")
	if err != nil {
		return nil, err
	}
	return db, nil
}

func checkSchemaRO(db *sql.DB, path string) error {
	var v string
	err := db.QueryRow("SELECT value FROM mundane_meta WHERE key='schema_version'").Scan(&v)
	if err != nil {
		return fmt.Errorf("schema_version not found in %s: %w", path, err)
	}
	if v != SchemaVersion {
		return &SchemaError{Path: path, Version: v}
	}
	return nil
}

func GetStatus(path string) (*Status, error) {
	db, err := openRO(path)
	if err != nil {
		return nil, err
	}
	defer db.Close()
	if err := checkSchemaRO(db, path); err != nil {
		return nil, err
	}
	st := &Status{Path: path}
	_ = db.QueryRow("SELECT value FROM mundane_meta WHERE key='task_id'").Scan(&st.TaskID)
	_ = db.QueryRow("SELECT value FROM mundane_meta WHERE key='created_at'").Scan(&st.CreatedAt)
	rows, err := db.Query("SELECT status, COUNT(*) FROM mundane_steps GROUP BY status")
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	for rows.Next() {
		var s string
		var c int
		if err := rows.Scan(&s, &c); err != nil {
			return nil, err
		}
		switch s {
		case StatusDone:
			st.Done = c
		case StatusPending:
			st.Pending = c
		case StatusFailed:
			st.Failed = c
		}
		st.Total += c
	}
	return st, nil
}

func GetSteps(path string) ([]InspectStepRow, error) {
	db, err := openRO(path)
	if err != nil {
		return nil, err
	}
	defer db.Close()
	if err := checkSchemaRO(db, path); err != nil {
		return nil, err
	}
	rows, err := db.Query(
		"SELECT id, name, kind, encoding, status, started_at, " +
			"COALESCE(finished_at, ''), COALESCE(error, '') " +
			"FROM mundane_steps ORDER BY id",
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []InspectStepRow
	for rows.Next() {
		var r InspectStepRow
		if err := rows.Scan(&r.ID, &r.Name, &r.Kind, &r.Encoding, &r.Status, &r.StartedAt, &r.FinishedAt, &r.Error); err != nil {
			return nil, err
		}
		out = append(out, r)
	}
	return out, rows.Err()
}

// GetResult returns the cached payload for `name`. The returned bytes are the
// emit-ready form: text/b64-decoded bytes for text/b64 steps; raw JSON text
// for json steps; ASCII integer for epoch rows.
func GetResult(path, name string) (encoding string, payload []byte, err error) {
	db, err := openRO(path)
	if err != nil {
		return "", nil, err
	}
	defer db.Close()
	if err := checkSchemaRO(db, path); err != nil {
		return "", nil, err
	}
	var enc, status string
	var raw []byte
	row := db.QueryRow(
		"SELECT encoding, result, status FROM mundane_steps WHERE name = ?",
		name,
	)
	if err := row.Scan(&enc, &raw, &status); err != nil {
		if err == sql.ErrNoRows {
			return "", nil, fmt.Errorf("no step named %q", name)
		}
		return "", nil, err
	}
	if status != StatusDone {
		return "", nil, fmt.Errorf("step %q is %s, not done", name, status)
	}
	return enc, raw, nil
}
